### PR TITLE
Add a BandwidthLogging transport wrapper

### LINKED ISSUE #887: Add a Transport wrapper that measures average bandwidth
We can copy the `bandwidth.rs` file from https://github.com/paritytech/substrate/pull/1551 and modify the license.

Once done and once a new version of libp2p is released, we can update Substrate to use it.

### PR DESCRIPTION
**Pull request by tomaka**
Close #887

### CODE REVIEW

File: src/bandwidth.rs [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+pub struct BandwidthLogging<TInner> {
+    inner: TInner,
+    sinks: Arc<BandwidthSinks>,
+}
+
+impl<TInner> BandwidthLogging<TInner> {
+    /// Creates a new `BandwidthLogging` around the transport.
+    #[inline]
+    pub fn new(inner: TInner, period: Duration) -> (Self, Arc<BandwidthSinks>) {
+        let mut period_seconds = cmp::min(period.as_secs(), u64::from(u32::max_value())) as u32;
```
Comment (twittner) [inline src/bandwidth.rs]: How about just capping the period to one POSIX day?

```suggestion
        let mut period_seconds = cmp::min(period.as_secs(), 86400) as u32;
```

File: src/bandwidth.rs [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+        if let Some(last) = self.bytes.last_mut() {
+            *last = last.saturating_add(bytes as u64);
+        }
+    }
+
+    /// Updates the state of the `BandwidthSink` so that the last element of `bytes` contains the
+    /// current second.
+    fn update(&mut self) {
+        let current_second = current_second();
+        for _ in self.latest_update .. current_second {
```
Comment (twittner) [inline src/bandwidth.rs]: Maybe we should limit the range to `self.bytes.len()` to avoid unnecessary loop iterations?

**Comment by tomaka** [comment]
I've fixed a bug and made some minor changes which could warrant a new review.

**Comment by tomaka** [comment]
However the changes are not large and quite uncontroversial, so if there's no new review by tomorrow I'll just merge.

**Pull request merged**